File: frontend/src/services/browser.ts
import { Events } from "@wailsio/runtime";
import { BrowserService } from "../../bindings/pi-desk/internal/appservice";
import type { BrowserStatus } from "../../bindings/pi-desk/internal/domain";

export interface BrowserEvent {
  type: "frame" | "attached" | "detached" | "navigated";
  sequence: number;
  dataB64?: string;
  cssWidth?: number;
  cssHeight?: number;
  url?: string;
  title?: string;
  error?: string;
}

export const browserService = {
  start(): Promise<BrowserStatus> {
    return BrowserService.Start();
  },
  openUrl(url: string): Promise<BrowserStatus> {
    return BrowserService.OpenURL({ url });
  },
  stop(): Promise<void> {
    return BrowserService.Stop();
  },
  click(x: number, y: number, button?: string, clickCount?: number): Promise<void> {
    return BrowserService.Click({ x, y, button, clickCount });
  },
  wheel(x: number, y: number, deltaX: number, deltaY: number): Promise<void> {
    return BrowserService.Scroll({ x, y, deltaX, deltaY });
  },
  key(key: string, modifiers: number): Promise<void> {
    return BrowserService.Key({ key, modifiers });
  },
  type(text: string): Promise<void> {
    return BrowserService.Type({ text });
  },
};

export function onBrowserEvent(callback: (event: BrowserEvent) => void): () => void {
  return Events.On("browser:event", (event) => callback(event.data as BrowserEvent));
}

File: internal/domain/browser.go
package domain

type BrowserStatus struct {
	Attached   bool   `json:"attached"`
	URL        string `json:"url,omitempty"`
	Title      string `json:"title,omitempty"`
	ProfileDir string `json:"profileDir,omitempty"`
}

type BrowserOpenURLRequest struct {
	URL string `json:"url"`
}

type BrowserClickRequest struct {
	X          float64 `json:"x"`
	Y          float64 `json:"y"`
	Button     string  `json:"button,omitempty"`
	ClickCount int     `json:"clickCount,omitempty"`
}

type BrowserWheelRequest struct {
	X      float64 `json:"x"`
	Y      float64 `json:"y"`
	DeltaX float64 `json:"deltaX"`
	DeltaY float64 `json:"deltaY"`
}

type BrowserKeyRequest struct {
	Key       string `json:"key"`
	Modifiers int    `json:"modifiers,omitempty"`
}

type BrowserTextInputRequest struct {
	Text string `json:"text"`
}

type BrowserEvent struct {
	Type      string `json:"type"`
	Sequence  uint64 `json:"sequence"`
	DataB64   string `json:"dataB64,omitempty"`
	CssWidth  int    `json:"cssWidth,omitempty"`
	CssHeight int    `json:"cssHeight,omitempty"`
	URL       string `json:"url,omitempty"`
	Title     string `json:"title,omitempty"`
	Error     string `json:"error,omitempty"`
}
